### PR TITLE
1745 refactor provider on v1 api

### PR DESCRIPTION
**Pull request by davidgisbey**
### Context
The V1 API is currently not cycle aware. 

### Changes proposed in this pull request

- exposes recruitment cycle in v1 provider serialiser
- creates a scope for all courses in a specific cycle
- refactors the provider controller to use the recruitment year in the url if there
- if not defaults to current recruitment cycle

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/controllers/api/v1/providers_controller.rb [hunk 1]
```diff
@@ -16,10 +16,7 @@ class ProvidersController < API::V1::ApplicationController
       # - clock drift between servers
       def index
         # only return 2019 courses until rollover is supported
-        if params[:recruitment_year].present? && params[:recruitment_year] != '2019'
-          render json: [], status: :not_found
-          return
-        end
+        build_recruitment_cycle
```
Comment (defong) [inline app/controllers/api/v1/providers_controller.rb]: i thought you wanted the `before_action`

also the comments above is kinda not making sense any morish

File: app/controllers/api/v1/providers_controller.rb [hunk 2]
```diff
@@ -16,10 +18,6 @@ class ProvidersController < API::V1::ApplicationController
       # - clock drift between servers
       def index
         # only return 2019 courses until rollover is supported
```
Comment (defong) [inline app/controllers/api/v1/providers_controller.rb]: rogue comment?

File: app/models/provider.rb [hunk 3]
```diff
@@ -74,6 +74,10 @@ class Provider < ApplicationRecord
 
   scope :in_order, -> { order(:provider_name) }
 
+  scope :by_recruitment_cycle, ->(recruitment_year) {
+    joins(:recruitment_cycle).merge(RecruitmentCycle.where(year: recruitment_year))
```
Comment (defong) [inline app/models/provider.rb]: is there a need for a join?

given that a provider belongs to a recruitment year?

File: spec/requests/api/v1/provider_spec.rb [hunk 4]
```diff
@@ -275,6 +279,42 @@ def get_provider_codes_from_body(body)
...
+
+          it 'defaults to the current cycle when year' do
+            returned_provider_codes = get_provider_codes_from_body(response.body)
+
+            expect(returned_provider_codes).not_to include provider2.provider_code
+            expect(returned_provider_codes).to include provider.provider_code
+          end
+        end
+        context 'with a future recruitment cycle specified in the route' do
+          let(:get_index) { get '/api/v1/2020/providers', headers: { 'HTTP_AUTHORIZATION' => credentials } }
```
Comment (defong) [inline spec/requests/api/v1/provider_spec.rb]: wondering should this be condense with 

for current
`'/api/v1/providers'`
`'/api/v1/#{current}/providers'`

for future
`'/api/v1/#{future}/providers'`

File: app/controllers/api/v1/providers_controller.rb (hunk 1 above)
Comment (davidgisbey) [reply app/controllers/api/v1/providers_controller.rb]: Done 👍

File: app/models/provider.rb (hunk 3 above)
Comment (davidgisbey) [reply app/models/provider.rb]: Done 👍

File: spec/requests/api/v1/provider_spec.rb (hunk 4 above)
Comment (davidgisbey) [reply spec/requests/api/v1/provider_spec.rb]: Seems like it's longer than just having 2020 though right?

**Comment by davidgisbey** [comment]
[[S] Change apiv1 course to return 2020 recruitment cycle](https://trello.com/c/ss65HD8u/1647-s-change-apiv1-course-to-return-2020-recruitment-cycle)

**Comment by davidgisbey** [comment]
[Refactor V1 Provider controller/serializers to reflect the new data model](https://trello.com/c/K4L4hzC1/1745-refactor-v1-provider-controller-serializers-to-reflect-the-new-data-model)

**Comment by davidgisbey** [comment]
[Allow recruitment cycle to be specified as a query parameter for backwards compatibility](https://trello.com/c/xgqaAf6E/1721-allow-recruitment-cycle-to-be-specified-as-a-query-parameter-for-backwards-compatibility)

**Pull request merged**